### PR TITLE
Add quotes to paths in MinGW build scripts (for paths with spaces)

### PR DESCRIPTION
**Pull request by Zapeth**

### CODE REVIEW

**Comment by cxd4** [comment]
The first and third commits are perfectly fine.  It is extremely difficult for me to want to merge the second.

I would have recommended organizing disputes and changes as big as those for VCH in separate pull requests.

**Comment by Zapeth** [comment]
Whoops, didn't mean to PR all commits at once (just the first), wanted to do the other two separately.

File: vu/select.c [hunk 1]
```diff
@@ -226,68 +226,46 @@ INLINE static void do_ch(pi16 VD, pi16 VS, pi16 VT)
...
+        {
+            ge[i] = (VS[i] >= VT[i]);
+            le[i] = (VT[i] < 0x0000);
+            cf_vce[i] = 0;
+
+            eq[i] = (VS[i] == VT[i]);
+            VC[i] = ge[i] ? VT[i] : VS[i];
+        }
+        cf_ne[i] = ~eq[i] & 1;
+    }
```
[the remark this answers is not in the document]
Comment (Zapeth) [reply vu/select.c]: As for the SIMD structure changes, I wasn't aware that a change like that could interfere with the compiler optmizations (I was actually wondering why you structured the rest of the code like that), thanks for the explanation.

I can try to restructure my changes accordingly, or we can further look into why your VCH implementation didn't produce correct results for the test rom (it does on cen64, so think its safe to say that the test rom works as intended).

[the remark this answers is not in the document]
Comment (Zapeth) [reply vu/select.c]: I don't know what source you used for your VCH implementation but I wrote mine according to the pseudocode of the instruction listed in the "N64 RSP Programmers Guide" doc.

I'm not sure if its safe to link directly to it here, but it should be among the first results if you do an online search for the quoted words.

**Comment by Zapeth** [comment]
I'm not quite sure how to properly submit PRs for each of these commits right now (I'm not that experienced with git/github in that regards).

If you want I can start from scratch with my fork (assuming thats still possible) and hold off pushing each commit until you accepted the according PR (I can omit the one for VCH or do it last for a separate review/discussion).
Alternatively I think its possible for you to cherry-pick the commits and merge them into your master branch? In that case feel free to do it for the commits that you approve and leave the rest open for the time being.

**Comment by LegendOfDragoon** [comment]
@Zapeth do you know what the inputs were, for the VRCPL test? I'm intrigued by your change.

Also for VCH, I don't see a functional difference for VCE. His original code essentially did the same thing already.

**Comment by cxd4** [comment]
And thank you @Zapeth likewise for bringing up the discussion of these points and code changes.

Git and GitHub indeed have quite a learning curve to them, but once you learn the counter-intuitive stuff, it becomes pretty hard to forget for some reason.  As I've recommended to @LegendOfDragoon on more than one occasion, probably the most rapid way to contribute over GitHub is to use multiple branches while keeping the `master` branch entirely untouched.

I'll attempt to go through example, but if you don't follow it right away it's fine.  Just do one PR at a time via your master branch and I'll merge one at a time.  Keep in mind that since you've already sent a pull request from your master branch, that integration became permanent on your local end, and since I did not merge it it means you would unfortunately have to redo the commits AFAIK.  Only commits by everyone else except yourself should be done to the `master` branch IMO.

So, example ... creating three different branches using Git command-line (I have no idea how to do it in GUIs if those are what you use though):
```sh
git checkout -b PR-for-adding-quotes
git checkout -b PR-for-fixing-VCH
git checkout -b PR-for-fixing-VRCPL-VRSQL
```
If you omit the `-b` flag it just switches to an existing branch by that name.  -b is something you pass for the very first time to create it in order for you to switch to it, and GitHub web interface lets you delete branches you no longer need (e.g. after a PR is merged).

Then instead of `git push origin master` like you always do by default when updating your master branch (which you should almost never do when contributing to someone else's repo), you would `git push origin PR-for-adding-quotes`, `git push origin PR-for-fixing-VCH`, `git push origin PR-for-fixing-VRCPL-VRSQL`, where each of those branches were committed to using the `git commit` command after switching to the appropriate one with `git checkout` before doing commits.

This way not only can you do several PRs simultaneously at one time without waiting for each one to be3 approved/rejected individually, but also in the case that one of the PRs was to be declined you would instantly recover because you did not do it from your master branch but a disposable topic branch with a custom name.

tl;dr -- if it's hard to follow or seems too much work, don't really want to inconvenience anyone.  I can just accept one-at-a-time PRs from master if need be.

>Alternatively I think its possible for you to cherry-pick the commits and merge them into your master branch? In that case feel free to do it for the commits that you approve and leave the rest open for the time being.

I do know how to do that manually (commit your changes by hand myself), but I tend to hold that off for last because I'd rather not take credit for others' changes or contributions under my username if they committed it for me.

File: vu/select.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (cxd4) [reply vu/select.c]: For whatever it's worth I am familiar with the N64 RSP Programmers Guide (for better or worse).  I cannot recall if krom himself ever got a hold of that or if that is still something all the #n64dev'ers still keep withheld from him.

>As for the SIMD structure changes, I wasn't aware that a change like that could interfere with the compiler optmizations (I was actually wondering why you structured the rest of the code like that), thanks for the explanation.

Yeah it's very hard to read my VCH/VCL code because those vector ops are under the select and compare group.  Multiplies, adds, divides, logical...all pretty straightforward ops you don't need if/else to emulate.  Selects like VCH, VCL, VCR and friends...it's extremely hard to write them without using if/else statements, but I had no choice.

Maybe this could help you?  Here is the most recent version of my repository where VCH was written in normal, scalar code without all the vectors, if you want to compare against that:

https://github.com/cxd4/rsp/blob/0f98ed926f85bf777fd4ce9e478262955adf93b0/vu/vch.h

>Iirc the results of your VCH implementation produced 2's complement results for the failed tests when it should have been *(-1) results, as well as incorrect results for VC0 and VCE flags.

I did at some point go over the functions and avoid the -(negation) operator due to implementation-defined behavior on having one's-, two's- or other-complement bitwise negation for different systems.  Maybe I tried a little too hard, or maybe there is some discrepancy in the interface for comparing to what krom did (as LegendOfDragoon proposed the emulator running the plugins perhaps).

**Comment by LegendOfDragoon** [comment]
The only way for cxd4's VCH to be wrong is if the documentation cxd4 used was also wrong. Interestingly, i was reviewing div instructions last night (haven't looked at it in ages lol). I don't believe it's always supposed to sign extend like what you suggest, but I'd like to see the inputs where the results didn't match.

>Thats interesting, what version of pj64 did you use for testing?
2.4. I'm even more confused now. I disabled 32 bit engine and now i see some of them fail. Here's what i see with 32bit engine disabled 
image
When 32bit engine was enabled, VRCPL & VRSQL passed everything. As you can see in my screenshot, it's a little hard to read the text lol. This was with cxd4's code before merging your changes.

**Comment by cxd4** [comment]
You're welcome to re-attempt VCH contribution, just with as few line changes or restructures as possible for easier analysis and verification on my end.  Maybe you find that there is more than one thing wrong with my VCH function or more than one domain of inputs that can be used to break how I implemented it vs. how it should be on the hardware, but try to do a PR that only fixes a single thing with a function that nasty, complicated and sensitive.

Not sure if it helps you play around at all, but I pasted a version of my current VCH function with all the inner auto-vectorizing `for` loops removed.  Instead of saying `for (i = 0; i < N; i++)` a million times, here is the version with saying that just once and encasing the entire procedure within that loop.  This may be easier for you to test with or manipulate; I just never write the vector ops emulation that way because it cannot be vectorized to SIMD as easily as doing a for loop for each and every single step.

https://gist.github.com/cxd4/52b8a521ec7eb71fd5b6e91e447bbc36/revisions

Oh right, wouldn't want to forget to tell you.  See the #ifdef _DEBUG, #elif 0, #else directives in VCH I did as well?  Those are for debugging.  See if you can reduce the number of simultaneous corrections you need to make to my function to get it to match krom's test just by switching to the _DEBUG or #elif 0 block alternatives, because that IS tested code that I know is stable but just do not use simply because of being prone to scalar code generation instead of SIMD optimizations.

>At least I think the changes for the latter are correct, I didn't spot any mentioning in the docs that DivIn should be unsigned (and the passing of the test cases seems to back that up).

I see the contrary.  I think the docs seem to resemble my DivIn definition more than the one without the unsigned cast you removed, but there is still sufficient evidence to merge that PR in addition, at least preliminarily for now.  See comments and discussion I left on the RCPL/RSQL PR for details.

**Comment by Zapeth** [comment]
I looked again that the results of the failed test cases for your VCH implementation and the "fix" would actually be quite simple - just do `VC[i] += sn[i]` after you force sn[i] to be either 1 or 0.

However to be honest, I'm not really sure about the purpose of your code in the rest of that area (specifically everything in relation with `diff[i]`), so I wouldn't really consider this a solution unless you say its valid.
Also this might still produce incorrect results where your corner case hack would apply (what cases would that be anyway?)
Not to mention the weird emulator inconsistencies reported by @LegendOfDragoon , though I have no idea what could be responsible for an off-by-one inconsistency like that.

**Comment by cxd4** [comment]
>Not to mention the weird emulator inconsistencies reported by @LegendOfDragoon , though I have no idea what could be responsible for an off-by-one inconsistency like that.

I would say that the fact that he reports some differences to be caused by the usage of the "Assume 32-bit" setting is somewhat telling enough.  I'm not terribly surprised since many inaccuracies or "bugs" in emulators are subtle enough to not amount to what ultimately is put up on screen by the video plugin.

>I looked again at the results of the failed test cases for your VCH implementation and the "fix" would actually be quite simple: just do VC[i] += sn[i] after you force sn[i] to be either 1 or 0.

How complete is the test ROM you are using for VCH?  I feel like the tests he is running which you are helping to get my plugin to pass may be limited to just a few cases and not a full range of combinations of inputs for VS[], VT[]/VC, and the carry flags.

For example if I did VC[i] += sn[i] (or actually VC[i] -= sn[i] maybe since I originally forced sn[i] to be -1 or 0, not +1 or 0...the former being more directly analogous to SSE2 code generation), that would potentially negate the conditional subtraction based on the corner-case hack done via VC[i] -= sn[i] & cch[i].  It is likely your change would pass hardware tests that my plugin is failing but also create regressions in which the majority of games would have graphical issues.

Ultimately I would propose testing the plugin by building it with the scripts which you helped fix with quotes (though I never thought about that I guess...on WinXP I know "Documents and Settings" is a likely candidate for file paths having spaces in them when compiling) with one or two games.  Usually if you change the RSP in a way that breaks one game badly, it's highly unlikely that which game you test matters--i.e. it'll usually break close to all of them.  So testing before committing is easy more-or-less.

>However to be honest, I'm not really sure about the purpose of your code in the rest of that area (specifically everything in relation with diff[i]),

The idea behind diff[] was to also vectorize comparisons between VS[i] and VT[i], not to mention the possible inclusion of the carry flag added to VS[i] or VT[i] throughout various phases.

So if you have `if` statements like `if (VS[i] < VT[i])`, you would eliminate scalar code generation and optimize to SIMD by changing to `if (VS[i] - VT[i] < 0)`, and you can remove the `if` statement eventually because (x < 0) means the most-significant bit, the sign bit, is set in _x_, assuming a 2's complement CPU.

Again, more reasons why my function is hard to read and deal with just to run as efficiently as possible or near close to it.  I did link you to an older snapshot of my repository long before I vectorized all that stuff if you wanted to try building my plugin from a commit that old, since that function was what I based off of in creating the vectorized version.  I would then re-derive what I did to make the current version of the function and apply your correction.

Maybe there are other RSP plugins you can test with PeterLemon's test ROM on VCH.  Their VCH() functions are far more easy to read than my own at this stage, so if you alternatively decided to find out what was needed to fix VCH for those RSP plugins that would also give me clues on the accuracy of my own implementation.

So there are various ways to go about it, just unfortunately none of which are straightforward.

>Also this might still produce incorrect results where your corner case hack would apply (what cases would that be anyway?)

The corner-case hack is relevant only if VT[i] == -32768 because my formula takes the absolute value of VT[i] for some reason I cannot offhand remember right away.

If you look at the manual's definition for how VCH sets the _NOTEQUAL_ carry flags, that should correspond to where I set eq[].  Then at the very end of the function I set the _NOTEQUAL_ flags to eq[i] ^ 1 in that loop.

**Comment by cxd4** [comment]
Hm...try setting cch[i] to 0 in all cases and remove my corner-case hack entirely to see if that passes the hardware tests you are running?  After all, it shouldn't be needed unless VT[i] in particular is INT16_MIN (which is a "corner-case" value after all).

**Comment by Zapeth** [comment]
> How complete is the test ROM you are using for VCH? I feel like the tests he is running which you are helping to get my plugin to pass may be limited to just a few cases and not a full range of combinations of inputs for VS[], VT[]/VC, and the carry flags.

Thats very likely, but due to the way these roms are structured its unfortunately relatively tedious to add more test cases to these roms right now (not to mention that I wouldn't be able to test them on real hardware anyway).
All I can say is that cen64 passes these failed tests for VCH (as does pj64 with its 32-bit engine apparently), so I don't think that there is something wrong with the tests or even the rom itself.

As for which tests fail (same for pj64 in default mode), its the first and third of the VCH tests. The results of VD are off-by-one for those tests, status flags seem to be correct though (except for VCE when testing in pj64, but I suspect thats an emulator specific issue, as I mentioned earlier).
For completness sake here are the values of the first test with your current implementation in mupen64:
```
VS = 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
VT = 00000011 00002233 00004455 00006677 ffff8899 ffffaabb ffffccdd ffffeeff
VD = 00000000 00000000 00000000 00000000 00007766 00005544 00003322 00001100
VCO = fff0
VCC = f0f0
VCE = 00
```
And here for the third test (if you know why VCE is ff here, feel free to explain it to me because  as far as I know, the flags in VCE are supposed to indicate whether abs(x[i]) == abs(y[i]), or am I mistaken?):
```
VS = 00000011 00002233 00004455 00006677 ffff8899 ffffaabb ffffccdd ffffeeff
VT = ffffffee ffffddcc ffffbbaa ffff9988 00007766 00005544 00003322 00001100
VD = 00000011 00002233 00004455 00006677 ffff8899 ffffaabb ffffccdd ffffeeff
VCO = 00ff
VCC = 0fff
VCE = ff
```

> Hm...try setting cch[i] to 0 in all cases and remove my corner-case hack entirely to see if that passes the hardware tests you are running?

No changes, though cch[i] has been zero for all cases so far anyway.

**Comment by cxd4** [comment]
Yeah not sure where the whole $vce flags issue is coming into the discussion from for everyone but doesn't seem to be any issues with those flags that I can see in the test ROM.

Just downloaded his test and ran it in Mupen64 64-bit under Linux:
https://n64.gitlab.io/cpu/rcp/rsp/vch-fail.gif

That'll remind me to make further modifications to my fork of angrylion's plugin since turning off the DAC filters doesn't seem to show the entire picture very well with a 32-bit RGBA frame buffer, which mind you is a rare design choice.  Had made many improvements since the last release of the plugin to EmuTalk as far as showing the full frame buffer and taking the full screenshots (like having the full 320x240 video capture and not just 320x237), but for crazy demoscene ROMs like this I had to dig up my PixD utility to get a legible screenshot of krom's test results.

>All I can say is that cen64 passes these failed tests for VCH (as does pj64 with its 32-bit engine apparently), so I don't think that there is something wrong with the tests or even the rom itself.

CEN64's not exactly known for being accurate (at least, as accurate as what you'd expect).  However long ago before I did any SIMD optimizations to the RSP, MarathonMan originally did begin his implementation of RSP op-codes from what I had then; that was long before we both rewrote everything from the ground up time and time again.  Especially after I got on his bad side. :P

Still trying to figure out how to read the screens...pretty counter-intuitive.  I'll see what I can do to make or break it some more.

**Comment by cxd4** [comment]
So have you guys (upon running the test in my RSP plugin) have just been hitting Enter, clicking OK or whatever through all the warning message spamming when trying to run krom's test ROMs with the plugin, without telling me it's been spamming them?

Cause I was beginning to think I'd never find a demo ROM that had SP task type = 0 and was not purposely throwing that error box up from within the plugin as a notification that something which should never have happened, happened, if anyone thought that.

**Comment by Zapeth** [comment]
Yeah, pretty much. I couldn't really find any information about these task types (apart from the ones defined in the n64 os lib) and since the roms are homebrew anyway I figured it wouldn't be worth mentioning.
Not to mention that the message only gets printed to console in the background for mupen64, so its not really interrupting anything.

**Comment by LegendOfDragoon** [comment]
>So have you guys (upon running the test in my RSP plugin) have just been hitting Enter

Lol yea, forgot to mention that.

>Still trying to figure out how to read the screens...pretty counter-intuitive.

I felt the same way and honestly gave up :smile: .

**Comment by cxd4** [comment]
I believe everything is taken care of now as far as the VCH, VRCPL accuracy comments in this thread.

Zapeth and all others please feel free to check with other hardware tests and test ROMs to find any remaining or future discrepancies in RSP output.

>I felt the same way and honestly gave up.

I think it would have been best (both for completeness and readability) if he somehow just generated a table of VD outputs from VS and VT taken from a real N64, looped through them and compared them to the results in the running session.  Then display only the inputs and outputs for which the outputs don't match the pre-recorded correct answer.

**Comment by LegendOfDragoon** [comment]
> Then display only the inputs and outputs for which the outputs don't match the pre-recorded correct answer.

SOmehow, I never thought of this for a test rom lol. Good idea :smile: . Anyway, I recently found a case where https://github.com/cxd4/rsp/blob/master/vu/divide.c#L1083 works differently. I only tested VRCPL, but if divin = 0x7FFFFFFF the result is different. Not sure why i didn't write down the result, but oh well lol. I honestly gave up on trying to a good alternative to the branching, so I just decided to use conditional move.

**Comment by cxd4** [comment]
Thinking about it, I can believe your observation without actually testing the code.

If DivIn = +0x7FFFFFFF as you say, then we have an input that is greater than -32768 (obviously).

The line you linked to hovers `data ^= (s32)(data + 32768) >> 31;`.  That works as intended as an alternative to `data ^= (data < -32768) ? ~0 : 0` only up until the hazard of signed overflow in 32-bit storage.

My intended range of values was to know if `data` is negative and small, but by staticizing with the code you linked to we are adding 32768 and checking the sign bit afterwards, and +0x7FFFFFFF + 32768 overflows to a negative number by signed wraparound.

So going off of blind hypothesis here the "quick fix" to what you say would I think be to try changing `(s32)(data + 32768)` to `(s32)((s64)data + 32768)` so that the 64-bit storage prevents 32-bit wraparound.

>I honestly gave up on trying to a good alternative to the branching, so I just decided to use conditional move.

... but having said all that, I'd prefer this.

Conditional moves shouldn't be that big of a deal with ops that are already scalar in nature.  VRCP[L/H] are all scalar operations tagged as vector under a technicality, so I'm not too pressed about optimizing them for branch elimination due to the internal scalar and "branchy" nature of how they actually operate anyway.

When I was folding branches and making code static everywhere I was really trying to focus on the real vector math because not only is SIMD an easy way to speed up the simple ops like VAND, VOR, VXOR and friends but also it is more "accurate" with how the real hardware executes, instead of iterating each 16-bit result 8 times.

Going a step beyond that and micro-optimizing more direct ops like VRCPL for branch elimination was just my younger past self being audacious and excited.  It is wise to see the pattern of problems arising from all of this as premature optimization, to take a step back, and to code for accuracy and self-documenting readability.

So the branch could certainly be kept, but I'd rather delete it down to the `#else` directive below what you linked and just use the conditional move.  I'll think about it more when I wake up.

>SOmehow, I never thought of this for a test rom lol. Good idea.

Well my key problem was having to use the word "somehow".

It is a perfect idea indeed but only once you can get the C code for the array of real N64 outputs of VS and VT included into your program.  Since jotting down all the outputs of VS and VT combinations by hand is pretty much out of the question we need to write a N64 program that will store the results to a text file for us, and I don't know if that's possible because the N64 doesn't have much of a file system.  Maybe use 64drive or try to write the file to flash RAM...dunno.

**Comment by theboy181** [comment]
If you need test on flash cart let me know.

**Comment by cxd4** [comment]
Thought of one solution...write all the VD[] results of VS[] and VT[] from -32768:+32767 to RDRAM, then use a GameShark, 64Drive or whatever to dump RDRAM and get it onto your PC.  Then write a program which parses the RDRAM and creates a C text source file out of it with each output of VD stored into an array indexable by VS and VT.

Then you can write a N64 homebrew executing a RSP operation and comparing the results to that huge gigantic look-up table of pre-recorded N64 results you extracted.

**Comment by cxd4** [comment]
@LegendOfDragoon I think the latest commit at https://github.com/cxd4/rsp/commit/e3c7f46090d4f5f231d45a45833047fc8f9184d7 addresses what you said except that I disabled the very same optimization I fixed out of relative indifference.

But in theory if you re-enable the optimization by changing the `#if (... && (0)` to !(0) I think it should pass the tests you were doing, just very busy and don't have time to look at it in-depth or write test code.

**Comment by cxd4** [comment]
Flash save chip would only be 128 KiB which is 128 * 1024 = 2 * 65536, and each 16-bit vector result requires 2 bytes.  So you get enough space to store 65,536 results, when you actually need 65,536 _squared_.  So if you dumped f(VS, VT) where VS = 0, VT = 0..65535, you would have to update VS 65,535 more times and dump the flash RAM save chip 65,536 times total to get the full results.

RDRAM would barely be big enough to fit the table, if you use as much of the 2-GiB limit as possible.

>The Vector Multiply Accumulate instructions have way too many inputs.

Instead of storing each 16-bit result from VD into RDRAM for the MAC operations, I would store each 48-bit accumulator.  Then flash the RDRAM back and forth 3 times instead of 1 time to make up for the fact you require 3 times as much memory, as that will never fit into a single 2 GB RDRAM dump.

Leave the accumulator at some constant value instead of varying that across runs of the opcode I would say.  The purpose of making a test ROM for VMA[C/D]* shouldn't be to test your emulator's ability to add to a 48-bit integer; it should be to test for the correctness of the multiplication.  Adding to the accumulator is a fairly mundane step which is hard to get wrong, so force the 48 bits to some kind of perhaps special but constant value before every run of the multiply instructions.  0 if you want to keep the test as small as possible.

>But something like VRCP or VRSQ would be very doable.

I guess the hard part would be testing every possible input of VS and VT, while VRCP only uses one vector instead of two which drastically reduces the amount of RDRAM you'd need to use for storage.

Actually, VS and VT follow the commutative mathematical property in most op-codes I would say, so 65536 * 65536 is actually much greater than what you'd really need.  e.g. if you test (VS, VT) = 0, 32767...you don't need to store the result of (VS, VT) = 32767, 0.

Except maybe for stuff using the carry flags under select.  Can't swear that the order is flexible there.  Not to mention you'd naturally want to store the results of the carry flags $vco, $vcc and $vce as well for all VS and VT inputs.  Probably make a separate test ROM for that...one testing VCH/VCL's carry flag outputs, (which are 1 or 2 bytes each anyway), the other testing VD.

>I'm really just doing all this for practice though. Optimizing the RDP would be far more worthwhile if I was just purely interested in results.

A worthwhile pursuit but once which tends to leave me feeling kind of introverted.

In general nobody really cares to interact with things I do for practice, although I've done a lot of personal work and research projects solely for that reason when I want people to leave me alone :) .  I like the feeling of being able to apply what I do in real world problems.

Kinda like filling up my paper notebook with floods of ink dedicated to deriving the cubic formula for trigonometry during my English teacher's boring lectures...fun and feels great, just doesn't apply to any matters in real life or get any communication going.

Occasionally micro-optimizations and coding for efficiency also give me opportunities to apply rather than practice, just not as often as coding for security, solidity or artistic elegance.  Given special contexts, it's almost more fun, tempting or even challenging to code an algorithm to be as slow as possible (if I may exaggerate), rather than as efficient as possible.

So yeah, in my younger anti-pragmatist days, I was perfectly content with coding just for practice and not results.  Gave it a few more years though...I don't have the mental stability or autism to crunch away at a bunch of abstract numbers and algorithms and not get irritated.  I get bored a lot faster than I used to at that stuff.

**Comment by Zapeth** [comment]
> Zapeth and all others please feel free to check with other hardware tests and test ROMs to find any remaining or future discrepancies in RSP output.

As a matter of fact, there are actually quite a few other tests that fail. However as far as I can tell pretty much all of them are either unofficial opcodes or opcodes that are used in an unofficial/unexpected way and the plugin logs issues accordingly for each of those (the only exception being RSPTransposeMatrixVMOV, though I suspect thats testing some unofficial/unexpected scenario as well).

Anyway, it looks like all the tests that I was hoping to see fixed are passing now in mupen64, so I think we can end this discussion here. If any more issues or problematic tests show up it would probably be best to open up a new issue for them (it wasn't really my intention to throw all of this discussion in one pull request, but I guess it ended up like that anyway).

As a last note from my side, thank you @cxd4 for taking the time to look into these issues and especially for the detailed and insightful responses throughout these discussions, its very appreciated.

**Comment by theboy181** [comment]
RSP developers come out of the woodworks when someone says something intelligent, so the thread getting derailed was inevitable.

**Comment by LegendOfDragoon** [comment]
>Leave the accumulator at some constant value instead of varying that across runs of the opcode I would say. 

You're right about that. It's much more reasonable to just focus on the multiplication part.

>, VS and VT follow the commutative mathematical property in most op-codes I would say, so 65536 * 65536 is actually much greater than what you'd really need

Indeed. I think you can reduce it by almost half, just by skipping the redundant numbers for a lot of these instructions. Still, i'd rather just narrow it down and test hundreds/ maybe thousands of cases, rather than millions/ billions. Although I think it would be neat if dozens of people each tested different cases.

>In general nobody really cares to interact with things I do for practice, although I've done a lot of personal work and research projects solely for that reason when I want people to leave me alone :) . I like the feeling of being able to apply what I do in real world problems.

I can definitely relate :smile: . Part of the reason i recently started reviewing the RSP again, was to see if I could apply knowledge/ skills I gained from doing other things.

>So yeah, in my younger anti-pragmatist days, I was perfectly content with coding just for practice and not results. Gave it a few more years though...I don't have the mental stability or autism to crunch away at a bunch of abstract numbers and algorithms and not get irritated. I get bored a lot faster than I used to at that stuff.

I'm generally content as long as I feel I have learned something worthwhile. I still get excited optimizing tiny stuff that don't make a significant impact on run-time :smile: . In many cases, my practice on minor optimizations came in handy when i attempted bigger optimizations later on.

**Comment by cxd4** [comment]
>As a matter of fact, there are actually quite a few other tests that fail. However as far as I can tell pretty much all of them are either unofficial/reserved opcodes or opcodes that are used in an unofficial/unexpected way and the plugin prints according warnings for each of those (the only exception being RSPTransposeMatrixVMOV, though I suspect thats testing some unofficial/unexpected scenario as well).

The newly opened discussions at #20, #21, #22, #23, and #24 (if I overlooked or missed anything let me know) I believe cover everything outside the range of the discussion you intended to open up.  If it wasn't for you I would never have bothered to clone his repository and test the full set of test ROMs krom did.  Indeed that wasn't intended to be all thrown into this one thread, but we don't get a whole lot of positive and constructive N64 activity going on and the gap in communications desensitized the issue of where to say what was said.

>Part of the reason i recently started reviewing the RSP again, was to see if I could apply knowledge/ skills I gained from doing other things.

If you're still going to school, look for ways to work your way up into the system with internships.  That honestly beats any sort of video gaming emulator project in this dimension you could ever come back to.

>In many cases, my practice on minor optimizations came in handy when i attempted bigger optimizations later on.

Micro-optimizations developing the foundation for major optimizations was one of the original starting ideas for me as well; just don't grow careless and forget that the opposite practicality often turns out to work at least as effectively if not more so:  Originating from the bigger picture and the larger scale goal of the project and working your way down into major and micro optimizations as warranted, rather than the inverse direction.

**Comment by LegendOfDragoon** [comment]
>That honestly beats any sort of video gaming emulator project in this dimension you could ever come back to.

I stuck with emulation because it was fun and, there was a good opportunity for learning how to optimize. But now, I'm starting to look into other things, like learning more libraries/APIs. Would be nice to be able to work on more types of projects. There are certainly better things to do than emulation. I just like trying stuff, when ideas pop up in my head :smile: .

>just don't grow careless and forget that the opposite practicality often turns out to work at least as effectively if not more so:

Indeed. I brainstorm a lot more these days, before working on things. I even get criticism for doing that, but I think it's effective :smile: . In the past, I tended to just focus on micro optimizations because I didn't know what else to do lol. Now that I have more experience, focusing on the bigger picture first, is more effective.
